### PR TITLE
Make OpenSSL build on Darwin

### PR DESCRIPTION
**Pull request by eschnett**

### CODE REVIEW

File: var/spack/packages/openssl/package.py [hunk 1]
```diff
@@ -17,12 +17,24 @@ class Openssl(Package):
...
+            perl("./Configure",
+                 "--prefix=%s" % prefix,
+                 "--openssldir=%s/etc/openssl" % prefix,
+                 "zlib",
+                 "no-krb5",
+                 "shared",
+                 "darwin64-x86_64-cc")
+            perl('-pi', '-e', 's/-arch x86_64//g', 'Makefile')
+        else:
+            exit(1)
```
Comment (tgamblin) [inline var/spack/packages/openssl/package.py]: this breaks OpenSSL for non-mac users.

**Comment by eschnett** [comment]
Corrected.

**Comment by eschnett** [comment]
Note to self: Use `filter_file(regex, repl, *filenames, **kwargs)` instead of `perl -pi -e`.

**Comment by mathstuf** [comment]
I had #189 which was sufficient. It just needs that `spec.satisfies("=darwin-x86_64")` condition in front of it.

**Comment by tgamblin** [comment]
So is there a need to have two different configure invocation formats?  Should both builds be using `Configure` or `config`?

**Comment by mathstuf** [comment]
I think that my branch with the environment setting inside of a `satisfies` conditional is the simplest version. The separate configure line certainly isn't *necessary* for a 64-bit OS X build, but OpenSSL is convoluted enough that I could see that as being another solution.

**Comment by eschnett** [comment]
According to OpenSSL's build instructions, Linux and Darwin should either call different configure scripts (as I did), or one should set `KERNEL_BITS`.

@mathstuf Did you try building with gcc? The call to `filter_file` from my patch is necessary if one wants to build with a compiler that is not Apple's `clang`.

**Comment by mathstuf** [comment]
No, I didn't try with gcc (I don't think it's installed, but it does have macports already).

**Comment by mathstuf** [comment]
`KERNEL_BITS` is also required on `vulcan`, a BG/P machine (ppc64).

**Comment by eschnett** [comment]
What is the architecture string for Vulcan? `ppc64`?

**Comment by mathstuf** [comment]
Yes, `ppc64`.

**Comment by mathstuf** [comment]
It appears that `KERNEL_BITS` should be set for any 64-bit target. Though why OpenSSL is saying "you know, you probably wanted this, but we're not going to do it and here's what you need to do to make it so" is…unfortunate. I'm guessing that 64-bit Linux x86_64 made enough noise to make it Do The Right Thing™ there, but everyone else is left with this unfortunate "solution".

**Comment by eschnett** [comment]
I combined and simplified the two patches.

File: var/spack/packages/openssl/package.py [hunk 2]
```diff
@@ -17,12 +17,18 @@ class Openssl(Package):
...
+            # where it happens automatically?)
+            env['KERNEL_BITS'] = '64'
         config = Executable("./config")
         config("--prefix=%s" % prefix,
-               "--openssldir=%s/etc/openssl" % prefix,
+               "--openssldir=%s" % join_path(prefix, 'etc', 'openssl'),
                "zlib",
                "no-krb5",
                "shared")
+        # Remove non-standard compiler options if present
```
Comment (mathstuf) [inline var/spack/packages/openssl/package.py]: Make a note that this is for GCC on OS X.

**Comment by mathstuf** [comment]
Otherwise, LGTM.

**Comment by eschnett** [comment]
Note added.

**Comment by tgamblin** [comment]
@eschnett, @mathstuf: Thanks!

**Pull request merged**